src/cwe_checker_lib/src/analysis/taint/state/memory_taint.rs

```diff
@@ -0,0 +1,212 @@
+//! Tracking of taint in memory.
+
+use crate::abstract_domain::{
+    AbstractIdentifier, DomainMap, MapMergeStrategy, MemRegion, UnionMergeStrategy,
+};
+
+use super::Taint;
+
+use std::collections::BTreeMap;
+
+/// Strategy for merging two memory taint states.
+///
+/// Essentially a [`UnionMergeStrategy`], i.e., the set of keys is the union of
+/// the individual key sets, but we do not use the merging provided by the
+/// [`MemRegion`] type on the intersection. Instead, we implement our own
+/// merging of `MemRegion<Taint>` in [`merge_memory_object_with_offset`].
+///
+/// [`UnionMergeStrategy`]: crate::abstract_domain::UnionMergeStrategy
+#[derive(Debug, PartialEq, Eq, Clone)]
+struct MemoryTaintMergeStrategy {
+    _private: (), // Marker to prevent instantiation
+}
+
+impl MapMergeStrategy<AbstractIdentifier, MemRegion<Taint>> for MemoryTaintMergeStrategy {
+    fn merge_map_with(
+        memory_taint: &mut BTreeMap<AbstractIdentifier, MemRegion<Taint>>,
+        other_memory_taint: &BTreeMap<AbstractIdentifier, MemRegion<Taint>>,
+    ) {
+        for (aid, other_memory_object) in other_memory_taint.iter() {
+            memory_taint
+                .entry(aid.clone())
+                .and_modify(|memory_object| {
+                    merge_memory_object_with_offset(memory_object, other_memory_object, 0);
+                })
+                .or_insert_with(|| other_memory_object.clone());
+        }
+    }
+}
+
+/// Represents our knowledge about taint in memory at a particular point in the
+/// program.
+pub type MemoryTaint = DomainMap<AbstractIdentifier, MemRegion<Taint>, UnionMergeStrategy>;
```

Comment:
But you defined the type `MemoryTaint` with `UnionMergeStrategy` instead of `MemoryTaintMergeStrategy`. So I think that the code does not use your intended implementation right now.